service/llm.py

```
@@ -1,16 +1,585 @@
+                    self.token_expiry = datetime.now() + self.token_lifetime
+                    logger.info("Token refreshed successfully")
+                    return
+            
+                except requests.exceptions.RequestException as e:
+                        logger.error(f"Attempt {attempt + 1}/{self.retry_attempts} failed: {str(e)}")
+                        if response and hasattr(response, 'text'):
+                            logger.error(f"Response content: {response.text}")
+                        if attempt < self.retry_attempts - 1:
+                            time.sleep(self.retry_delay)
+                        else:
+                            raise RuntimeError(f"Failed to refresh token after {self.retry_attempts} attempts")
```

Ruff (0.8.2)

Within an `except` clause, raise exceptions with `raise ... from err` or `raise ... from None` to distinguish them from errors in exception handling

(B904)